src/ralph/api/routers/statements.py

```diff
@@ -0,0 +1,244 @@
+            "multiple pages."
+            "NB: for internal use, not part of the LRS specification."
+        ),
+    ),
+):
+    """Fetch a single xAPI Statement or multiple xAPI Statements."""
+
+    # Create a list to aggregate all the enabled filters for the search
+    es_query_filters = []
+
+    if statementId:
+        es_query_filters += [{"term": {"_id": statementId}}]
+
+    if agent:
+        es_query_filters += [{"term": {"actor.account.name.keyword": agent}}]
+
+    if verb:
+        es_query_filters += [{"term": {"verb.id.keyword": verb}}]
+
+    if activity:
+        es_query_filters += [
+            {"term": {"object.objectType.keyword": "Activity"}},
+            {"term": {"object.id.keyword": activity}},
+        ]
+
+    if since:
+        es_query_filters += [{"range": {"timestamp": {"gt": since}}}]
+
+    if until:
+        es_query_filters += [{"range": {"timestamp": {"lte": until}}}]
+
+    if len(es_query_filters) > 0:
+        es_query = {"query": {"bool": {"filter": es_query_filters}}}
+    else:
+        es_query = {"query": {"match_all": {}}}
+
+    # Honor the "ascending" parameter, otherwise show most recent statements first
+    es_query.update(
+        {"sort": [{"timestamp": {"order": "asc" if ascending else "desc"}}]}
+    )
+
+    # Create a point-in-time or use the existing one to ensure consistency of search
+    # results over multiple pages.
+    if not pit_id:
+        # pylint: disable=unexpected-keyword-arg
+        pit_id = ES_CLIENT.open_point_in_time(index="statements", keep_alive="1m")["id"]
+    es_query.update(
+        {
+            "pit": {
+                "id": pit_id,
+                "keep_alive": "1m",  # extend duration of PIT whenever it is used
+            }
+        }
+    )
+
+    if search_after:
+        es_query.update({"search_after": search_after.split("|")})
+
+    # Make sure the limit does not go above max from settings
+    limit = min(limit, ES_MAX_SEARCH_HITS_COUNT)
+
+    # pylint: disable=unexpected-keyword-arg
+    es_response = ES_CLIENT.search(
+        body=es_query,
+        size=limit,
+    )
+
+    # Prepare the link to get the next page of requests, while preserving the
+    # consistency of search results.
+    # NB: There is an unhandled edge case where the total number of results is
+    # exactly a multiple of the "limit", in which case we'll offer an extra page
+    # with 0 results.
+    response = {}
+    if len(es_response["hits"]["hits"]) == limit:
+        # Search after relies on sorting info located in the last hit
+        last_hit_sort = [str(part) for part in es_response["hits"]["hits"][-1]["sort"]]
+        path = request.url.components.path
+        query = request.url.components.query
+        response.update(
+            {
+                "more": (
+                    f"{path}{query + '&' if query else '?'}pit_id={pit_id}"
+                    f"&search_after={'|'.join(last_hit_sort)}"
+                )
+            }
+        )
```

Comment:
I don't see pagination management in the response when the number of statements exceeds the limit?

Reply:
The number of statements cannot exceed limit by definition.

Either it is the same if there are more results than the `limit` number (because limit meant we only got `limit` results, eg `1000` results out of `10000000`), or it is smaller if there  are fewer than `limit` results (eg `limit` is `1000` and there are `300` results).

Reply:
That's also how I use `limit`. Therefore a page can only ever contain `limit` (eg. 1000) statements or less, never more.

Then I use this to decide if I should include `"more"` in the response:  if the number of results I'm getting is less than `limit`, that means I've reached the last page of results.

If there are 6450 results, the first 6 pages will have 1000 results, and the last one 450: when I get 450 results, I know I've reached the end of my pagination — and when I have a page with 1000 results, I'm assuming I get 1000 because I have more than 1000 results remaining. Of course, if there are exactly 6000 results this breaks down and I'll get 6 pages with 1000 results and one page with zero result.

Reply:
Oh, thanks for the clear explanation: I finally see the light!

I would have the opposite approach that adds the `more` field by default unless if `len(statements) < limit`, I find it easier to understand. WDYT?